src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,288 @@
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132601
+            var tokens = ComputeTokens(
+                data, ref lastLineNumber, ref lastStartCharacter, text.Lines, groupedSpans);
+
+            return new LSP.SemanticTokens { ResultId = (previousResultId + 1).ToString(), Data = tokens };
+        }
+
+        private static int[] ComputeTokens(
+            ArrayBuilder<int> data,
+            ref int lastLineNumber,
+            ref int lastStartCharacter,
+            TextLineCollection lines,
+            IEnumerable<IGrouping<TextSpan, ClassifiedSpan>> groupedSpans)
+        {
+            foreach (var span in groupedSpans)
+            {
+                ComputeNextToken(data, lines, ref lastLineNumber, ref lastStartCharacter, span);
+            }
+
+            return data.ToArray();
+        }
+
+        private static void ComputeNextToken(
+            ArrayBuilder<int> data,
+            TextLineCollection lines,
+            ref int lastLineNumber,
+            ref int lastStartCharacter,
+            IGrouping<TextSpan, ClassifiedSpan> span)
+        {
+            // Each semantic token is represented in LSP by five numbers:
+            //     1. Token line number, relative to the previous token
+            //     2. Token start character, relative to the previous token
+            //     3. Token length
+            //     4. Token type (index) - looked up in SemanticTokensLegend.tokenTypes
+            //     5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+
+            var textSpan = span.Key;
+            var linePosition = lines.GetLinePositionSpan(textSpan).Start;
+            var lineNumber = linePosition.Line;
+            var startCharacter = linePosition.Character;
+
+            // 1. Token line number, relative to the previous token
+            var deltaLine = lineNumber - lastLineNumber;
+            data.Add(deltaLine);
+
+            // 2. Token start character, relative to the previous token
+            // (Relative to 0 or the previous token’s start if they're on the same line)
+            if (lastLineNumber == lineNumber)
+            {
+                data.Add(startCharacter - lastStartCharacter);
+            }
+            else
+            {
+                data.Add(startCharacter);
+            }
+
+            // 3. Token length
+            data.Add(textSpan.Length);
+
+            var additiveResults = span.Where(s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
+
+            // 4. Token type - looked up in SemanticTokensLegend.tokenTypes.
+            var tokenTypeClassifiedSpan = span.Except(additiveResults).Single();
+            if (s_classificationTypeToSemanticTokenTypeMap.TryGetValue(tokenTypeClassifiedSpan.ClassificationType, out var tokenType))
+            {
+                var index = TokenTypes.IndexOf(t => t == tokenType);
```

Comment:
The mapping of token name to int seems like something that should be defined in the LSP protocol code, at least for the base LSP types.  Otherwise every server has to maintain this separately.  Almost seems like int Enum values would be perfect here.

cc @gundermanc